src/messages/LimitedQueue.hpp

```diff
@@ -2,296 +2,245 @@
+    [[nodiscard]] T back() const
     {
-        std::vector<T> acceptedItems;
+        return this->buffer_.back();
+    }
 
-        if (this->space() > 0)
-        {
-            std::lock_guard<std::mutex> lock(this->mutex_);
+    // Modifiers
 
-            // create new vector to clone chunks into
-            auto newChunks = std::make_shared<ChunkVector>();
+    // Clear the buffer
+    void clear()
+    {
+        std::unique_lock lock(this->mutex_);
 
-            newChunks->resize(this->chunks_->size());
+        this->buffer_.clear();
+    }
 
-            // copy chunks except for first one
-            for (size_t i = 1; i < this->chunks_->size(); i++)
-            {
-                newChunks->at(i) = this->chunks_->at(i);
-            }
+    /**
+     * @brief Push an item to the end of the queue
+     *
+     * @param item the item to push
+     * @param[out] deleted the item that was deleted
+     * @return true if an element was deleted to make room
+     */
+    bool pushBack(const T &item, T &deleted)
+    {
+        std::unique_lock lock(this->mutex_);
 
-            // create new chunk for the first one
-            size_t offset =
-                std::min(this->space(), static_cast<qsizetype>(items.size()));
-            auto newFirstChunk = std::make_shared<Chunk>();
-            newFirstChunk->resize(this->chunks_->front()->size() + offset);
+        bool full = this->buffer_.full();
+        if (full)
+        {
+            deleted = this->buffer_.front();
+        }
+        this->buffer_.push_back(item);
+        return full;
+    }
 
-            for (size_t i = 0; i < offset; i++)
-            {
-                newFirstChunk->at(i) = items[items.size() - offset + i];
-                acceptedItems.push_back(items[items.size() - offset + i]);
-            }
+    /**
+     * @brief Push an item to the end of the queue
+     *
+     * @param item the item to push
+     * @return true if an element was deleted to make room
+     */
+    bool pushBack(const T &item)
+    {
+        std::unique_lock lock(this->mutex_);
 
-            for (size_t i = 0; i < this->chunks_->at(0)->size(); i++)
-            {
-                newFirstChunk->at(i + offset) = this->chunks_->at(0)->at(i);
-            }
+        bool full = this->buffer_.full();
+        this->buffer_.push_back(item);
+        return full;
+    }
 
-            newChunks->at(0) = newFirstChunk;
+    /**
+     * @brief Push items into beginning of queue
+     *
+     * Items are inserted in reverse order.
+     * Items will only be inserted if they fit,
+     * meaning no elements can be deleted from using this function.
+     *
+     * @param items the vector of items to push
+     * @return vector of elements that were pushed
+     */
+    std::vector<T> pushFront(const std::vector<T> &items)
+    {
+        std::unique_lock lock(this->mutex_);
 
-            this->chunks_ = newChunks;
+        size_t numToPush = std::min(items.size(), this->space());
```

Comment:
```suggestion
        size_t numToPush = std::min(items.size(), this->limit() - this->buffer_.size());
```

if a shared lock is applied to `this->size()`, calling `space()` here could be problematic without a recursive lock (as `space()` calls `this->size()`), so this should avoid a self-deadlock scenario